Sources/SmithyCBOR/Writer/Writer.swift

```diff
@@ -0,0 +1,224 @@
+        case .dateTime, .httpDate:
+            let string = TimestampFormatter(format: format).string(from: value)
+            self.cborValue = .text(string)
+        }
+    }
+
+    public func write<T>(_ value: T?) throws where T: RawRepresentable, T.RawValue == Int {
+        if let rawValue = value?.rawValue {
+            try write(rawValue)
+        }
+    }
+
+    public func write<T>(_ value: T?) throws where T: RawRepresentable, T.RawValue == String {
+        if let rawValue = value?.rawValue {
+            try write(rawValue)
+        }
+    }
+
+    public func writeMap<T>(
+        _ value: [String: T]?,
+        valueWritingClosure: (T, Writer) throws -> Void,
+        keyNodeInfo: NodeInfo,
+        valueNodeInfo: NodeInfo,
+        isFlattened: Bool
+    ) throws {
+        guard let value else { return }
+        var map: [String: CBORType] = [:]
+
+        for (key, val) in value {
+            let writer = self[key]
+            try valueWritingClosure(val, writer)
+
+            // If the writer itself doesn't have a cborValue, build it from its children
+            if writer.cborValue == nil, !writer.children.isEmpty {
+                var childMap: [String: CBORType] = [:]
+                for child in writer.children {
+                    if let childCborValue = child.cborValue {
+                        childMap[child.nodeInfo] = childCborValue
+                    }
+                }
+                writer.cborValue = .map(childMap) // Construct the map for the writer
+            }
+
+            // Add to the parent map
+            if let cborValue = writer.cborValue {
+                map[key] = cborValue
+            }
+        }
+
+        // Assign the constructed map to the current writer
+        self.cborValue = .map(map)
+    }
+
+    public func writeList<T>(
+        _ value: [T]?,
+        memberWritingClosure: (T, Writer) throws -> Void,
+        memberNodeInfo: NodeInfo,
+        isFlattened: Bool
+    ) throws {
+        guard let value else { return }
+
+        var array: [CBORType] = []
+
+        for val in value {
+            // Create a child writer for each list element
+            let childWriter = Writer(nodeInfo: memberNodeInfo, parent: self)
+            try memberWritingClosure(val, childWriter)
+
+            // If the child writer has a cborValue, add it to the array
+            if let cborValue = childWriter.cborValue {
+                array.append(cborValue)
+            } else if !childWriter.children.isEmpty {
+                // If no cborValue but has children, create a map from its children
+                var childMap: [String: CBORType] = [:]
+                for child in childWriter.children {
+                    if let childCborValue = child.cborValue {
+                        childMap[child.nodeInfo] = childCborValue
+                    }
+                }
+                array.append(.map(childMap)) // Append the constructed map
```

Comment:
Same questions as ones for map writing above, except for list instead of map.

Reply:
let me know if I answered above -- in this case see: ```// Create a child writer for each list element
            let childWriter = Writer(nodeInfo: memberNodeInfo, parent: self)```